app/cli.py

```diff
@@ -0,0 +1,239 @@
+import copy
+import json
+import multiprocessing
+import os
+import shlex
+from pathlib import Path
+
+import click
+import yaml
+from click_dict_type import DictParamType
+from clouds.aws.aws_utils import set_and_verify_aws_credentials
+from ocp_utilities.utils import run_command
+
+
+def generate_cluster_dir_path(clusters, base_directory):
+    for _cluster in clusters:
+        cluster_name = _cluster["name"]
+        platform = _cluster["platform"]
+        _cluster["install-dir"] = os.path.join(base_directory, platform, cluster_name)
+    return clusters
+
+
+def verify_processes_passed(processes, action):
+    failed_processes = {}
+
+    for _proc in processes:
+        _proc.join()
+        if _proc.exitcode != 0:
+            failed_processes[_proc.name] = _proc.exitcode
+
+    if failed_processes:
+        click.echo(f"Some jobs failed to {action}: {failed_processes}\n")
+        raise click.Abort()
+
+
+def update_base_install_config(pull_secret_file):
+    base_install_config = get_base_install_config_data()
+    base_install_config["pullSecret"] = json.dumps(
+        get_pull_secret_data(pull_secret_file=pull_secret_file)
+    )
+    base_install_config["sshKey"] = get_local_ssh_key()
+    return base_install_config
+
+
+def download_openshift_install_binary(clusters, pull_secret_file):
+    versions = set()
+    for cluster in clusters:
+        versions.add(cluster["version"])
+
+    for version in versions:
+        binary_dir = os.path.join("/tmp", version)
+        openshift_install_str = "openshift-install"
```

Comment:
can be outside the loop

Reply:
How? we create folder for each version

Reply:
`openshift_install_str` var is constant to the loop